funnel/models/event.py

```diff
@@ -137,11 +232,39 @@ class SyncTicket(BaseMixin, db.Model):
     participant_id = db.Column(None, db.ForeignKey('participant.id'), nullable=False)
     participant = db.relationship(Participant, primaryjoin=participant_id == Participant.id,
         backref=db.backref('sync_tickets', cascade="all, delete-orphan"))
-    proposal_space_id = db.Column(db.Integer, db.ForeignKey('proposal_space.id'), nullable=False)
-    proposal_space = db.relationship(ProposalSpace,
-        backref=db.backref('sync_tickets', cascade='all, delete-orphan'))
     ticket_client_id = db.Column(db.Integer, db.ForeignKey('ticket_client.id'), nullable=False)
     ticket_client = db.relationship(TicketClient,
         backref=db.backref('sync_tickets', cascade='all, delete-orphan'))
+    __table_args__ = (db.UniqueConstraint('ticket_client_id', 'order_no', 'ticket_no'),)
+
+    previous_participant = None
+
+    @classmethod
+    def get(cls, ticket_client, order_no, ticket_no):
+        return cls.query.filter_by(ticket_client=ticket_client, order_no=order_no, ticket_no=ticket_no).one_or_none()
+
+    @classmethod
+    def upsert(cls, ticket_client, order_no, ticket_no, **fields):
+        """
+        Returns a tuple containing the upserted ticket, and the participant the ticket
+        was previously associated with or None if there was no earlier participant.
+        """
+        ticket = cls.get(ticket_client, order_no, ticket_no)
+        if ticket:
+            if ticket.participant is not fields.get('participant'):
+                # Transfer ticket
+                ticket.previous_participant = ticket.participant
```

Comment:
(Actually, it may not, since you're referring to the instance and not the class. Still, seems shady.)